### ISSUE TITLE
Force install Swift for TensorFlow on Google Colaboratory

### ISSUE THREAD

**Issue by RahulBhalley** [tensorflow/swift#621]
Hi @dan-zheng!

I hope you're doing well. S4TF was an amazing project! Must've not been archived btw. Only a giant organization could have made Swift relevant in Deep Learning space. I was working with _Metal Performance Shaders_ library from @apple and it's not a good API (not even documented, very buggy, and error-prone). You guys literally did fantastic job! 😃 I just wish someone, maybe @google itself, please revive this project. 

Anyway, I wanna use S4TF again in Colaboratory. Is there anyway I can **force install S4TF**? Or is there any other way on the Earth 🌎 I can use GPU (or TPU) using S4TF?

Regards
Rahul Bhalley

**Comment by BradLarson** [comment]
Hey, Rahul, thanks for the kind words. The project certainly was a lot of fun, and many components (differentiable Swift, Python interoperability, etc.) continue to live on and be useful in various domains.

Unfortunately, when it comes to Colab, the Swift Jupyter kernel currently requires some functionality that is not enabled in Colab by default. As a result, you aren't able to side-load Swift support into Colab in the same manner as you can Julia. It might be possible at some point in the future to overcome that restriction in the Swift Jupyter kernel, but that may take some work. Marc and others might be able to elaborate on what would be needed.

Local Swift Jupyter notebooks still work well, and give you full access to your local GPU on Linux. Google Cloud instances should also let you access GPU resources via command-line applications or hosted Jupyter notebooks. I believe they also should provide a means to work on TPUs via Swift for TensorFlow.

I hear where you're coming from on the Metal Performance Shaders, because I built a graph framework on them once that processed Caffe files and built a graph at runtime. They are great building block components, because of the speed they can deliver on macOS / iOS, but you really need to read the headers to get padding, byte-ordering, and so on correct. The new [Metal Performance Shaders Graph](https://developer.apple.com/documentation/metalperformanceshadersgraph) API does make it easier to build out graphs for training and inference.

**Comment by RahulBhalley** [comment]
Hi Brad, thanks for the information.

> Local Swift Jupyter notebooks still work well, and give you full access to your local GPU on Linux. Google Cloud instances should also let you access GPU resources via command-line applications or hosted Jupyter notebooks.

Thanks, I don't have a GPU myself and just want to use it on cloud for _free_. Btw can I install S4TF and run programs using `swift` CLI on Kaggle? I think I was able to once install S4TF on AWS's EC2 (but didn't try running it on any accelerators) when I was doing my startup around 2019 I think and I'm not sure if I can do that on Kaggle's GPU/TPU instances. Have you tried it and could it be possible to do that?

> I hear where you're coming from on the Metal Performance Shaders, because I built a graph framework on them once that processed Caffe files and built a graph at runtime. They are great building block components, because of the speed they can deliver on macOS / iOS, but you really need to read the headers to get padding, byte-ordering, and so on correct. The new Metal Performance Shaders Graph API does make it easier to build out graphs for training and inference.

Actually, I was referring to MPS Graph itself. It's error prone and not much documented. For example, consider the case of order of weight parameters' shape indices, some `format` argument in `MPSConvolutionDescriptor` I think (used to specify the order) for transpose convolution 2D doesn't align with the actual shape of weight parameter values whereas the `format` does respect the values it's taking in the case of convolution 2D op. 

The API design is not good. It's should be intuitive like S4TF and PyTorch where graph construction is abstracted away behind the scenes and is not in user's hands.

**Comment by philipturner** [comment]
I found the problem, and bypassed Google’s restriction on Swift on Colab :-)
